nsis/src/nsis/file/NsisExecutable.java

```diff
@@ -213,4 +235,24 @@ public InputStream getDecompressedInputStream() throws IOException {
 		return this.decompressionProvider.getDecompressedStream();
 	}
 
+	/**
+	 * Get the number of pages in the Nsis executable
+	 * 
+	 * @return the number of pages
+	 */
+	public int getNumPages() {
+		return this.pages.length;
+	}
+
+	/**
+	 * Returns the data structure of a Nsis page
+	 * 
+	 * @return a DataType object that represents one Nsis page
+	 * @throws DuplicateNameException
+	 * @throws IOException
+	 */
+	public DataType getPageDataType() throws DuplicateNameException, IOException {
+		return this.pages[0].toDataType();
```

Comment:
I think you need to pass in the index here so you get the right structure?

Reply:
It is weird because toDataType() is not a static function, and it can't be static because I need to override it. No matter the index, the data type will be the same as it is common for all instances of NsisPage. Should i still need to specify the index even though it is kind of weird?

Reply:
I'm not sure I follow what you are saying.

I can clarify my suggestion:
- remove the `dataType` paramter to `initPagesSection`, add the `ne` variable for the `NsisExecutable` object.
- inside `initPagesSection` in the for loop call `ne.getPageDataType(i)` to return `this.pages[i].toDataType()`

Otherwise I think you will always be returning the data of the page at offset 0, I think you need to return each page's data, because each page has different data?

Reply:
- remove the dataType paramter to initPagesSection, add the ne variable for the NsisExecutable object.
I agree

- inside initPagesSection in the for loop call ne.getPageDataType(i) to return this.pages[i].toDataType()

getDataType does not return the data of each page, but the `STRUCTURE` object so it always returns the same `STRUCTURE` no matter on what page instance we call it on. If I could, I would make `getDataType()` a static function (even `STRUCTURE` is static) but i need to override that function so it works with Ghidra. 

![image](https://user-images.githubusercontent.com/24403370/87681053-2cb93a80-c74c-11ea-88bc-65d013c956d4.png)
![image](https://user-images.githubusercontent.com/24403370/87681077-380c6600-c74c-11ea-8507-db4b79f93e63.png)

Reply:
Would there be a way to make this cleaner given that `toDataType()` (in `NsisPage`) and `getPageDataType()` (in `NsisExecutable`) can't be static methods?

Reply:
Thanks for clarifying, I didn't quite understand the code, to be honest. Let me think...

Reply:
The object that actually tells Ghidra what is the data is the `DecompressedInputStream` we retrieve from `NsisExecutable` in the `NsisLoader`. The memory blocks are initialized with that input stream and then `STRUCTURE` objects are applied on top of that data, so Ghidra knows how to present it to the user

Reply:
This might work:
https://stackoverflow.com/a/10279596

```java
public DataType getPageDataType() throws DuplicateNameException, IOException {
  return NsisPage.STRUCTURE;
}
```

Maybe we could then get rid of `getPageDataType()`?

Reply:
Oh yeah that's much better! I don't know why I didn't think of that, thank you